File: easy/201-300/226/go/main.go
// Copyright 2021 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

package main

import "fmt"

type Fraction struct {
	Numerator   int
	Denominator int
}

func (f *Fraction) String() string {
	return fmt.Sprintf("%d/%d", f.Numerator, f.Denominator)
}

func (f *Fraction) gcd(first, second int) int {
	var a, b int
	if first > second {
		a, b = first, second
	} else {
		a, b = second, first
	}
	for 0 < a%b {
		a, b = b, a%b
	}
	return b
}

func (f *Fraction) Reduce() *Fraction {
	gcd := f.gcd(f.Numerator, f.Denominator)
	f.Numerator /= gcd
	f.Denominator /= gcd
	return f
}

func (f *Fraction) Add(fraction *Fraction) *Fraction {
	f.Numerator = f.Numerator*fraction.Denominator + fraction.Numerator*f.Denominator
	f.Denominator *= fraction.Denominator
	return f.Reduce()
}

var zPrint = fmt.Println

func main() {
	_, _ = zPrint("hello world")
}

File: easy/201-300/226/rust/src/main.rs
// Copyright 2022 CJ Harries
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

use gcd::Gcd;

struct Fraction {
    numerator: u32,
    denominator: u32,
}

impl Fraction {
    fn new(numerator: u32, denominator: u32) -> Fraction {
        Fraction {
            numerator: numerator,
            denominator: denominator,
        }
    }

    fn add(&self, other: &Fraction) -> Fraction {
        let new_numerator = self.numerator * other.denominator + other.numerator * self.denominator;
        let new_denominator = self.denominator * other.denominator;
        let gcd = new_numerator.gcd(new_denominator);
        Fraction::new(new_numerator / gcd, new_denominator / gcd)
    }
}

#[cfg(not(tarpaulin_include))]
fn main() {
    println!("rad");
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn test_fraction_new() {
        assert_eq!(2, Fraction::new(2, 1).numerator);
        assert_eq!(3, Fraction::new(1, 3).denominator);
    }
}
